core/domain/question_services.py

```diff
@@ -201,3 +227,37 @@ def update_question(committer_id, question_id, change_list, commit_message):
     """
     question = apply_change_list(question_id, change_list)
     _save_question(committer_id, question, change_list, commit_message)
+
+
+def get_questions_batch(
+        collection_id, skill_ids, user_id, batch_size):
+    """Fetches a batch of questions for a user based on the provided
+    skill_ids, matching them with user skills.
+
+    Args:
+        collection_id: str. Id of the collection.
+        skill_ids: list(str). A list of skill ids.
+        user_id: str. Id of the user.
+        batch_size: int. The intended number of questions to be returned.
+
+    Returns:
+        list. A list of Question objects.
+    """
+    user_skill_ids = (
+        collection_services.get_acquired_skills_of_user_given_collection_id(
+            user_id, collection_id))
+    question_skill_ids = list(set(user_skill_ids) & set(skill_ids))
+    collection = collection_services.get_collection_by_id(collection_id)
+    question_ids = []
+    for skill_id in question_skill_ids:
+        if skill_id in collection.skills:
+            question_ids.extend(collection.skills[skill_id].question_ids)
+    unique_question_ids = list(set(question_ids))
+    if len(unique_question_ids) < batch_size:
+        batch_size = len(unique_question_ids)
+    random_question_ids = random.sample(unique_question_ids, batch_size)
+
+    questions_batch = []
+    for question_id in random_question_ids:
+        questions_batch.append(get_question_by_id(question_id))
```

Comment:
Same question -- this needs to be done using get multi.

Reply:
Updated.